Add tutorial on GCMC

lburth opened the pull request:
Fixes #4575 

Description of changes:
- Adds a tutorial, that teaches users how to simulate systems in the Grand-Canonical ensemble with the Grand-Canonical Monte Carlo.

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+   "source": [
+    "## Grand-Canonical Ensemble\n",
+    "The case of changing particle numbers is best described by the grand canonical ensemble <a href='#[1]'>[1]</a>.\n",
+    "In the grand canonical ensemble, the volume $V$, the temperature $T$ and the chemical potentials $\\mu_i$ of the exchangeable species $i$ rather than the particle numbers $N_i$ are fixed.The chemical potentials $\\mu_i$correspond to the free energy cost of adding a particle of type $i$ while keeping everything else fixed:\n",
```
missing space in last sentence

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "The grand canonical ensemble is thus also referred to as the $(\\mu, V, T)$-ensemble.\n",
+    "In the grand canonical ensemble, the probability for the system to contain $N_i$ particles of type $i$, i.e. a total of $N=\\sum_{i=1}^{N_\\mathrm{s}}N_i$ and to be in a specific microstate $\\xi\\in\\Gamma_{N}$ is given by the probability distribution\n",
+    "\\begin{align}\n",
+    "p^\\mathrm{G}\\left(\\left\\{N_i\\right\\},\\xi\\right) = \\frac{\\exp\\left(-\\beta \\left(H(\\xi)-\\sum_{i=1}^{N_\\mathrm{s}}\\mu_iN_i\\right)\\right)}{Z^\\mathrm{G}h^{3N}\\prod_{i=1}^{N_\\mathrm{s}} N_i!}.  \\quad \\tag{1}\n",
```
define N_s

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "\\end{align}\n",
+    "Here, $H$ denotes the Hamiltonian of the system with $N_i$ particles of type $i$ and $Z^\\mathrm{G}$ is the grand-canonical partition function which is given by\n",
+    "\\begin{align}\n",
+    "Z^\\mathrm{G}\\left(\\left\\{\\mu_i\\right\\},V,T\\right) = \\sum_{N_1=0}^{\\infty}...\\sum_{N_\\mathrm{s}=0}^{\\infty}Z\\left(\\left\\{N_i\\right\\},V,T\\right)\\exp\\left(\\beta \\sum_{i=1}^{N_\\mathrm{s}}\\mu_iN_i\\right).\n",
```
define Z

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "\\mu_i^{\\mathrm{sys}} = \\mu_i^{\\mathrm{res}}.\n",
+    "\\end{equation}\n",
+    "This condition is referred to as a chemical equilibrium between the two phases and is equivalent to a minimization of the free energy of the total system (system + reservoir).\n",
+    "In the case where one requires electroneutrality of the phases, due to the additional constraint, the electrochemical potential $\\overline{\\mu}_i=\\mu_i+z_ie\\psi$ rather than the chemical potential has to be equal."
```
define z, e, psi

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "\\frac{P_{i\\rightarrow j}^\\mathrm{acc}}{P_{j\\rightarrow i}^\\mathrm{acc}} = \\frac{p_j^\\mathrm{eq}}{p_i^\\mathrm{eq}}. \n",
+    "\\quad \\tag{6}\n",
+    "\\end{align}\n",
+    "It is easy to see that the acceptance probability \n",
```
avoid such phrases to not discourage beginners

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "\n",
+    "2. Insertion and Deletion moves that add or delete a single particle while keeping all other particle positions fixed.\n",
+    "\n",
+    "In the GCMC method the insertion and deletion moves are implemented in the following fashion. First of all it is determined randomly with equal probability if an insertion or deletion move is performed. In the case of an insertion the additional particle is placed at a uniformly chosen random position in the simulation box. Then proposed new state is accepted according to the criterion\n",
```
missing "the" in last sentence

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "The case of changing particle numbers is best described by the grand canonical ensemble <a href='#[1]'>[1]</a>.\n",
+    "In the grand canonical ensemble, the volume $V$, the temperature $T$ and the chemical potentials $\\mu_i$ of the exchangeable species $i$ rather than the particle numbers $N_i$ are fixed.The chemical potentials $\\mu_i$correspond to the free energy cost of adding a particle of type $i$ while keeping everything else fixed:\n",
+    "\\begin{align}\n",
+    "\\mu_i = \\left(\\frac{\\partial F}{\\partial N_i}\\right)_{T, V, \\left\\{N_j\\right\\}_{j\\neq i}}.\n",
```
define F

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    " \\begin{align}\n",
+    "p^\\mathrm{eq}_{N} \\propto \\exp\\left(-\\beta U(\\mathbf{q})  \\right ) \\prod_i \\frac{1}{N_i!}\\left(\\frac{V}{\\Lambda_i^3}\\right)^{N_i}\\times \\exp \\left (\\beta \\mu_i N_i\\right)\n",
+    "\\end{align}\n",
+    "This simplification is valid if we are only interested in observables which do not depend on the momenta."
```
it's not really a simplification (in the sense of an approximation).

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "Here we want to investigate a polyelectrolyte solution that is coupled to a reservoir containing a monovalent salt, as shown schematically in figure 1. The small salt ions (X$^+$ and X$^-$) can be exchanged between the two phases while the polyelectrolyte is confined to the system. Experimentally such a setup could be realized by seperating the two phases with a semi-permeable membrane. Simi ...
+    "\n",
+    "Due to the macroscopic electroneutrality of both phases the salt ions obey an electrochemical equilibrium, i. e. there is a difference in the electrostatic potential between the two phases $$\\mu_i^{\\mathrm{sys}}+z_ie\\psi^{\\mathrm{Donnan}} = \\mu_i^{\\mathrm{res}}.$$\n",
+    "This effect is called the Donnan effect <a href='#[3]'>[3]</a> and leads to an unequal partioning of salt between the reservoir and the system. The partitioning can be quantified using the partition coeffiecents $\\xi_i \\equiv c^{\\mathrm{sys}}_i/c^{\\mathrm{res}}_i$. Defining the \"universal\" partion coefficient,\n",
```
define c

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "This effect is called the Donnan effect <a href='#[3]'>[3]</a> and leads to an unequal partioning of salt between the reservoir and the system. The partitioning can be quantified using the partition coeffiecents $\\xi_i \\equiv c^{\\mathrm{sys}}_i/c^{\\mathrm{res}}_i$. Defining the \"universal\" partion coefficient,\n",
+    "\\begin{align*}\n",
+    "     \\xi\\equiv\\left\\{\n",
+    "    \\begin{array}{ll}\n",
```
why do we have two different definitions?

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "    \\xi_-  & \\text{for negative ions,}\\\\\n",
+    "    \\end{array}\\right.\n",
+    "\\end{align*}\n",
+    "one can obtain the following solution for an ideal system:\n",
```
define ideal

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "\\end{align*}\n",
+    "one can obtain the following solution for an ideal system:\n",
+    "\\begin{align}\n",
+    "     \\xi^{\\text{Donnan}} = -\\frac{c_{\\text{M}^-}^{\\text{sys}}}{2c_{\\text{X}^+,\\text{X}^-}^{\\text{res}}} + \\sqrt{\\left(\\frac{c_{\\text{M}^-}^{\\text{sys}}}{2c_{\\text{X}^+,\\text{X}^-}^{\\text{res}}}\\right)^2+1}.\n",
```
maybe a plot would be nice if it's not too much work

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "from espressomd import electrostatics\n",
+    "from espressomd import reaction_methods\n",
+    "from espressomd import polymer\n",
+    "from espressomd.interactions import *\n",
```
avoid star-imports

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+   "outputs": [],
+   "source": [
+    "##### Units\n",
+    "sigma = 3.55e-10 # Sigma in SI units\n",
```
where does this specific sigma come from?

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "##### Units\n",
+    "sigma = 3.55e-10 # Sigma in SI units\n",
+    "avo = 6.022e+23 # Avogadro's number in SI units\n",
+    "pref = 1/(10**3 * avo * sigma**3) # Prefactor to mol/L\n",
```
could we use the pint package here instead of manual unit conversions?

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Setup of the simulation\n",
```
basically same header as previous section

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "\n",
+    "##### Weeks-Chandler-Andersen interaction\n",
+    "LJ_EPSILON = 1.0\n",
+    "LJ_SIGMA = 1.0\n",
```
how do these parameters relate to the SI-parameters introduced above?

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "print(\"Created system class\")\n",
+    "\n",
+    "#### Add the FENE interaction\n",
+    "fene = FeneBond(k=30, d_r_max=1.5, r_0=0.0)\n",
```
why is the bond length 0?

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "        previous_part = part\n",
+    "\n",
+    "##### Add the particles\n",
+    "N_plus = n_salt + N_CHAINS*N_MONOMERS_PER_CHAIN\n",
```
mention electroneutrality constraint and where the counterions come from

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "##### Add the interactions\n",
+    "for i in types:\n",
+    "    for j in types:\n",
+    "        system.non_bonded_inter[types[i], types[j]].lennard_jones.set_params(epsilon=LJ_EPSILON, sigma=LJ_SIGMA, cutoff=LJ_SIGMA * 2**(1.0/6.0), shift=\"auto\")\n",
```
you could use the built-in WCA interaction

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "        types[\"Xminus\"]: charges[\"Xminus\"] \n",
+    "    }\n",
+    ")\n",
+    "RE.set_non_interacting_type(type = len(types)+1)\n",
```
what does this line do?

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 50/50 [00:04<00:00, 11.72it/s]"
```
remove output from tutorial before commiting/pushing

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "    K_XX =salt_concentration ** 2 * np.exp(excess_chemical_potential_monovalent_pairs_in_bulk(salt_concentration))\n",
+    "    RE.change_reaction_constant(gamma= K_XX,reaction_id = 0)\n",
+    "    for i in tqdm(range(warmup_loops),bar_format='{l_bar}{bar:60}{r_bar}{bar:-60b}'):\n",
+    "        system.integrator.run(1000)\n",
```
maybe give these numbers as predefined variables

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+   },
+   "source": [
+    "**Exercise**\n",
+    "* Plot the universal partition coeffiecients obtained by measuring the partition coefficient $\\xi_-$ and $\\xi_+$ in the simulation against $c_{\\mathrm{M}^-}/(2c_{\\mathrm{salt}}^{\\mathrm{res}})$\n",
```
not sure plotting is a meaningful exercise

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+   "source": [
+    "**Exercise**\n",
+    "* Plot the universal partition coeffiecients obtained by measuring the partition coefficient $\\xi_-$ and $\\xi_+$ in the simulation against $c_{\\mathrm{M}^-}/(2c_{\\mathrm{salt}}^{\\mathrm{res}})$\n",
+    "* Interpret the deviation of the simulation results from the ideal prediction."
```
there should definitely be a solution to this question

christophlohrmann commented on doc/tutorials/grand_canonical_monte_carlo/grand_canonical_monte_carlo.ipynb:
```diff
@@ -0,0 +1,888 @@
+    "<figcaption>Fig. 1: Schematic representation of the considered two phase setup.</figcaption>\n",
+    "</center>\n",
+    "</figure>\n",
+    "Here we want to investigate a polyelectrolyte solution that is coupled to a reservoir containing a monovalent salt, as shown schematically in figure 1. The small salt ions (X$^+$ and X$^-$) can be exchanged between the two phases while the polyelectrolyte is confined to the system. Experimentally such a setup could be realized by seperating the two phases with a semi-permeable membrane. Simi ...
```
maybe add a sentence on the relevance/usefulness of this system

christophlohrmann reviewed (changes requested):
This is a very well written tutorial. I have a bunch of minor comments, the only thing major is that after the last plot, the physics should be discussed in more detail to interpret the main result.
Otherwise: Thanks a lot, well done

lburth commented:
Thank you for your review and your annotations. I have now worked in your annotations and added an interpretation of the final plot. In addition I have changed the sample sizes a bit to speed up the runtimes.

christophlohrmann commented:
Looks really nice now, thanks for the work you put in.
For me, the solutions to the exercises are shown when I open the notebook. Please check if it looks the same for you and make the default not-shown if needed.
Otherwise this can be merged

lburth commented:
> Looks really nice now, thanks for the work you put in. For me, the solutions to the exercises are shown when I open the notebook. Please check if it looks the same for you and make the default not-shown if needed. Otherwise this can be merged
Thanks for the review. Now the default is that the solutions are hidden.

lburth committed:
Add tutorial on GCMC

jngrad committed:
Modify GCMC tutorial

Reduce sampling times. Improve testing. Redesign plots.
